fido/Crypto/Fido2/Attestation/FidoU2F.hs

```diff
@@ -0,0 +1,88 @@
+import Data.X509 (Certificate (certPubKey), PubKey (PubKeyEC), PubKeyEC (PubKeyEC_Named, pubkeyEC_name, pubkeyEC_pub), SerializedPoint (SerializedPoint), getCertificate)
+import qualified Data.X509 as X509
+import Text.ParserCombinators.ReadPrec (step)
+
+-- https://www.w3.org/TR/webauthn-2/#sctn-android-key-attestation
+verify :: Stmt -> AuthenticatorData -> Digest SHA256 -> Either AttestationError AttestedCredentialData
+verify Stmt {sig, attCert} AuthenticatorData {rpIdHash, attestedCredentialData} clientDataHash = do
+  -- 1. Verify that attStmt is valid CBOR conforming to the syntax defined above
+  -- and perform CBOR decoding on it to extract the contained fields.
+  -- NOTE: The validity of the data is already checked during decoding.
+
+  -- 2.a Check that x5c has exactly one element and let attCert be that element.
+  -- NOTE: This has already been done during decoding
+
+  -- 2.b Let certificate public key be the public key conveyed by attCert. If
+  -- certificate public key is not an Elliptic Curve (EC) public key over the
+  -- P-256 curve, terminate this algorithm and return an appropriate error.
+  let certPubKey' = certPubKey $ getCertificate attCert
+  case certPubKey' of
+    -- TODO: Will we only get named curves?
+    (PubKeyEC PubKeyEC_Named {pubkeyEC_name = SEC_p256r1}) -> pure ()
+    _ -> Left U2FNoECKeyinCertificate
+
+  -- 3. Extract the claimed rpIdHash from authenticatorData, and the claimed
+  -- credentialId and credentialPublicKey from
+  -- authenticatorData.attestedCredentialData.
+  credData@AttestedCredentialData {credentialId, credentialPublicKey} <- maybe (Left CredentialDataMissing) pure attestedCredentialData
+
+  -- 4. Convert the COSE_KEY formatted credentialPublicKey (see Section 7 of
+  -- [RFC8152]) to Raw ANSI X9.62 public key format (see ALG_KEY_ECC_X962_RAW in
+  -- Section 3.6.2 Public Key Representation Formats of [FIDO-Registry]).
+
+  -- 4.a Let x be the value corresponding to the "-2" key (representing x
+  -- coordinate) in credentialPublicKey, and confirm its size to be of 32 bytes.
+  -- If size differs or "-2" key is not found, terminate this algorithm and
+  -- return an appropriate error.
+  -- 4.b Let y be the value corresponding to the "-3" key (representing y
+  -- coordinate) in credentialPublicKey, and confirm its size to be of 32 bytes.
+  -- If size differs or "-3" key is not found, terminate this algorithm and
+  -- return an appropriate error.
+  -- NOTE: The decoding already happened in the decoding step
+  -- TODO: Should we check if the length is 32 bytes?
+  (x, y) <- case credentialPublicKey of
+    ECDSAPublicKey (ECDSAKey _ (Point x y)) -> pure (x, y)
+    _ -> Left U2FNoECKeyInAttestedCredentialData
+
+  let xb = i2osp x
+      yb = i2osp y
+  unless (BS.length xb == 32 && BS.length yb == 32) $ Left U2FUnexpectedCoordinateLength
+
+  -- 4.c Let publicKeyU2F be the concatenation 0x04 || x || y.
+  let publicKeyU2F = i2osp 0x04 <> xb <> yb
```

Comment:
We could probably remove i2osp so we don't need to call that function here. With `BS.singleton (4 :: Word8)` perhaps?

Reply:
Sounds like a good idea